docs/src/algorithmic_differentiation.md

```diff
@@ -0,0 +1,451 @@
+We refer readers to the first few minutes of the first lecture mentioned above for a more careful explanation.
+
+_**Vector-to-Vector Functions**_
+
+The generalisation of this to Euclidean space should be familiar: if ``f : \RR^P \to \RR^Q`` is differentiable at a point ``x \in \RR^P``, then the derivative of ``f`` at ``x`` is given by the Jacobian matrix at ``x``, denoted ``J[x] \in \RR^{Q \times P}``, such that
+```math
+\text{d}f = J[x] \, \text{d}x .
+```
+
+It is possible to stop here, as all the functions we shall need to consider can in principle be written as functions on some subset ``\RR^P``.
+
+However, when we consider differentiating computer programmes, we will have to deal with complicated nested data structures, e.g. `struct`s inside `Tuple`s inside `Vector`s etc.
+While all of these data structures _can_ be mapped onto a flat vector in order to make sense of the Jacobian of a computer programme, this becomes very inconvenient very quickly.
+To see the problem, consider the Julia function whose input is of type `Tuple{Tuple{Float64, Vector{Float64}}, Vector{Float64}, Float64}` and whose output is of type `Tuple{Vector{Float64}, Float64}`.
+What kind of object might be use to represent the derivative of a function mapping between these two spaces?
+We certainly _can_ treat these as structured "view" into a "flat" `Vector{Float64}`s, and then define a Jacobian, but actually _finding_ this mapping is a tedious exercise even if it quite obviously exists.
+
+Similarly, while "vector-Jacobian" products are usually used to explain reverse-mode AD, a more general formulation of the derivative is used all the time -- the matrix calculus discussed by [giles2008extended](@cite) and [minka2000old](@cite) (to name a couple) make use of a generalised form of derivative in order to work with functions which map to and from matrices (despite slight differences in naming conventions from text to text).
+
+Consequently, it will be much easier to avoid these kinds of "flattening" operations wherever possible.
+In order to do so, we make use of a generalised notion of the derivative.
+
+_**Functions Between More General Spaces**_
+
+In order to avoid the difficulties described above, we consider we consider functions ``f : \mathcal{X} \to \mathcal{Y}``, where ``\mathcal{X}`` and ``\mathcal{Y}`` are _finite_ dimensional real Hilbert spaces (read: finite-dimensional vector space with an inner product, and real-valued scalars).
+This definition includes functions to / from ``\RR``, ``\RR^D``, but also real-valued matrices.
+Furthermore, we shall see later how we can model all sorts of structured representations of data directly as such spaces.
+
+For such spaces, the derivative of ``f`` at ``x \in \mathcal{X}`` is the linear operator (read: linear function) ``D f [x] : \mathcal{X} \to \mathcal{Y}`` satisfying
```

Comment:
It took me a while to parse this correctly. I think two things threw me off a bit:
1) It wasn't obvious to me that I can consider dx to be an element of `\mathcal{X}`, considering it's infinitesimal and gradients I think are elements of the dual space.
2) There are multiple layers of functions here, where `D` takes the function `f` as an argument, `D f` takes the point `x` as an argument, and `D f [x]` takes the infinitesimal d x as an argument.

Because of the above two, for a moment I thought the `\mathcal{X} \to \mathcal{Y}` part was referring to the `[x]` argument of `D f`, and this got me confused.

Reply:
Okay. I've clarified the fact that `D f [x]` is a function, and that I denote its application to an argument as `D f [x] (\dot{x})` sooner in the text.